src/java.base/share/classes/java/lang/String.java

```diff
@@ -2431,8 +2431,55 @@ public int indexOf(int ch) {
      *          if the character does not occur.
      */
     public int indexOf(int ch, int fromIndex) {
-        return isLatin1() ? StringLatin1.indexOf(value, ch, fromIndex)
-                          : StringUTF16.indexOf(value, ch, fromIndex);
+        return indexOf(ch, fromIndex, length());
+    }
+
+    /**
+     * Returns the index within this string of the first occurrence of the
+     * specified character, starting the search at {@code fromIndex} and
+     * stopping before {@code toIndex}.
+     *
+     * <p>If a character with value {@code ch} occurs in the
+     * character sequence represented by this {@code String}
+     * object at an index no smaller than {@code fromIndex} but smaller than
+     * {@code toIndex}, then
+     * the index of the first such occurrence is returned. For values
+     * of {@code ch} in the range from 0 to 0xFFFF (inclusive),
+     * this is the smallest value <i>k</i> such that:
+     * <blockquote><pre>
+     * (this.charAt(<i>k</i>) == ch) &amp;&amp; (fromIndex &lt;= <i>k</i> &lt; toIndex)
+     * </pre></blockquote>
+     * is true. For other values of {@code ch}, it is the
+     * smallest value <i>k</i> such that:
+     * <blockquote><pre>
+     * (this.codePointAt(<i>k</i>) == ch) &amp;&amp; (fromIndex &lt;= <i>k</i> &lt; toIndex)
+     * </pre></blockquote>
+     * is true. In either case, if no such character occurs in this
+     * string at or after position {@code fromIndex} and before position
+     * {@code toIndex}, then {@code -1} is returned.
+     *
+     * <p>There are no restrictions on the value of {@code fromIndex} and
+     * {@code toIndex}. Negative values have the same effect as it they were zero.
+     * Values greater than the length of this string have the same effect
+     * as if they were equal to the length of this string.
+     *
+     * <p>As consequence of these rules, if {@code fromIndex} is greater than
+     * or equal to {@code toIndex}, then {@code -1} is returned.
```

Comment:
Are there other examples in String where the equivalent of fromIndex > toIndex doesn't throw?

Reply:
I simply extrapolated the behavior from `indexOf(int ch,int fromIndex), which has a similar note:
```
     * There is no restriction on the value of {@code fromIndex}. If it
     * is negative, it has the same effect as if it were zero: this entire
     * string may be searched. If it is greater than the length of this
     * string, it has the same effect as if it were equal to the length of
     * this string: {@code -1} is returned.
```

Reply:
This is the behavior of the underlying implementation as well.

Reply:
Personally I would expect `string.substring(fromIndex, toIndex).indexOf(ch)` to behave isomorphically to `string.indexOf(ch, fromIndex, toIndex)`

Reply:
Then you would also expect `string.substring(fromIndex).indexOf(ch)` to behave isomorphically to `string.indexOf(ch, fromIndex)` in current releases, right?
It does not.

Reply:
(I guess you are referring to the fact that `substring()` can throw.)

Reply:
@rgiulietti That's surprising, but if it is as you say, then the indexOf with or without the endIndex should behave identically and not throw.

Reply:
@viktorklang-ora I agree that the behavior w.r.t. out-of-range indices is at odd with the `substring()` family, but this inconsistency is already present since the dawn of time ;-)

Reply:
@AlanBateman I can propose variants of forward-searching _some-prefix_`IndexOf()` which are consistent with `substring()` in their handling of the indices, but I'd then rather prefer to file another JBS issue and prepare another PR.

The proposed 3 parameters variant of `indexOf()` _is_ the real "primitive" of the family. The other ones are mere shortcuts.

Reply:
Parking this PR while exploration is done into other options is okay.

Reply:
@AlanBateman Both from an `@apiNote` and from an implementation perspective, the checked _some-prefix_`IndexOf()`variants would better make use of the public 3 params `indexOf()` API point proposed here, rather than relying on `StringLatin1` and `StringUTF16` internals.

Thus, parking this PR would mean that the other PR could not refer to the API point from here until this one is integrated. 

Perhaps I'm missing something.
Is it perhaps better to add the checked variants directly in this PR?

Reply:
The suggestion is to write down some alternatives to the 3-arg indexOf method and work through some examples to see if they are less error prone. It might be that you decide to move forward with the current proposal or it might be that you decide a different method would help avoid some common mistakes.